runtime/io/common/src/aws/smithy/kotlin/runtime/io/RealSdkByteChannel.kt

```diff
@@ -0,0 +1,212 @@
+        }
+    }
+
+    private suspend fun awaitBytesToWrite(requested: Int) {
+        while (availableForWrite < requested && !isClosedForWrite) {
+            if (!tryFlush()) {
+                slot.sleep { availableForWrite < requested && !isClosedForWrite }
+            }
+        }
+    }
+
+    private suspend fun awaitFreeSpace() {
+        flush()
+        awaitBytesToWrite(1)
+        ensureNotClosed()
+    }
+
+    override suspend fun read(sink: SdkBuffer, limit: Long): Long {
+        require(limit >= 0L) { "Read limit must be >= 0, was $limit" }
+        ensureNotFailed()
+        if (isClosedForRead) return -1L
+        if (limit == 0L) return 0L
+
+        return reading {
+            if (availableForRead == 0) {
+                awaitBytesToRead(1)
+                ensureNotFailed()
+                // closed while suspended
+                if (isClosedForRead) return -1L
+            }
+
+            val rc = minOf(availableForRead.toLong(), limit)
+
+            synchronized(lock) {
+                sink.write(buffer, rc)
+            }
+
+            afterRead(rc.toInt())
+
+            rc
+        }
+    }
+
+    private inline fun reading(block: () -> Long): Long = try {
+        check(_readInProgress.compareAndSet(false, true)) { "Read operation already in progress" }
+        block()
+    } finally {
+        _readInProgress.compareAndSet(true, false)
+    }
+
+    override suspend fun write(source: SdkBuffer, byteCount: Long) {
+        ensureNotClosed()
+        if (byteCount == 0L) return
+
+        writing {
+            var remaining = byteCount
+
+            while (remaining > 0) {
+                if (availableForWrite == 0) {
+                    awaitFreeSpace()
+                }
+
+                val wc = minOf(availableForWrite.toLong(), remaining)
+
+                synchronized(lock) {
+                    buffer.write(source, wc)
+                }
+
+                afterWrite(wc.toInt())
+                remaining -= wc
+            }
+        }
+    }
+
+    private inline fun writing(block: () -> Unit) =
+        try {
+            check(_writeInProgress.compareAndSet(false, true)) { "Write operation already in progress" }
+            block()
+        } finally {
+            _writeInProgress.compareAndSet(true, false)
+        }
+
+    // read side only
+    private fun ensureNotFailed() {
+        closedCause?.let { throw it }
+    }
+
+    // write side only
+    private fun ensureNotClosed() {
+        if (isClosedForWrite) {
+            throw closedCause ?: ClosedWriteChannelException("Channel $this is already closed")
+        }
+    }
+
+    private fun afterWrite(size: Int) {
+        capacity.completeWrite(size)
+        _totalBytesWritten.plusAssign(size.toLong())
+
+        if (autoFlush || availableForWrite == 0) {
+            flush()
+        }
+    }
+
+    private fun afterRead(size: Int) {
+        capacity.completeRead(size)
+        slot.resume()
+    }
```

Comment:
**Suggestion**: If there are things which _must_ occur after reading/writing, I'd consider making them part of the `reading`/`writing` methods.

Reply:
I don't recall if there was a specific reason to not include it there but I'll take a look and comment or move things around as appropriate